keps/draft-20180426-security-conformance.md

```diff
@@ -0,0 +1,682 @@
+
+#### Run as root
+
+  *Conformance Profile: Isolated*
+
+  Containers in Pod created without explicit RBAC associated with the Pod service account,
+  should run as non-root user.
+
+  _Note: this may change if a better isolation runtime (e.g. VM) is used for the container._
+
+###### Test Recommendation
+
+  This test requires a container image with the programs (e.g. `id`) to get current UID and GID.
+
+  The test code consumes Kubernetes API as a serviceaccount/user with privileges:
+
+  - Create/Delete a namespace
+  - Create a ReplicaSet in the new namespace
+
+
+  Test setup
+
+  - Create a new namespace
+
+
+  Test steps
+
+  - Create a ReplicaSet
+  - Poll the ReplicaSet and the namespace, and a new Pod should be created
+  - Check the output of the Pod (can be log) and make sure UID and GID are not zero
+
+
+  Test teardown
+
+  - Delete the namespace
+
+#### Network Isolation
+
+  *Conformance Profile: Isolated*
+
+  Ingress/Egress are blocked for all cross-namespace traffic for all Pods by default,
+  except the Egress to access `kube-dns`.
+  Namespaces `kube-system` and those not selected are not touched.
+
+###### Test Recommendation
+
+  This test requires a container image with the programs (e.g. `socat`) to communicate over network.
+  Each container will have a listening port open to receive messages and also a process periodically sending messages to an remote address.
+  The received messages will be printed on stdout.
+  The container image will auto perform a test to connect to kubelet read-only port.
+
+  The test code consumes Kubernetes API as a serviceaccount/user with privileges:
+
+  - Create/Delete a namespace
+  - Create Pods in the new namespace
+
+
+  Test setup
+
+  - Create two namespaces, ns1 and ns2
+
+
+  Test steps
+
+  - Create a Pod in ns1 to connect to kubelet read-only port
+  - Check the output to make sure the connection should fail;
+  - Create two Pods in ns1
+  - Check the output to make sure they can communicate with each other
+  - Create two Pods, one in ns1, and the other in ns2
+  - Check the output to make sure they can't talk to each other.
+
+
+  Test teardown
+
+  - Delete the namespaces
+
+#### API server configuration
+
+  *Conformance Profile: Essential*
+
+  The following are required:
+
+  - All communication over mTLS/TLS
+    - apiserver talks to etcd via HTTPS
+    - controller-manager, scheduler talks to apiserver via HTTPS
+
+
+  The following should be disabled:
+
+  - insecure listening port
+    threat: the attack may access the insecure port to by-pass authn/authz
+  - basic-auth
+    threat: weak authentication
+  - token-auth-file
+    threat: weak authentication
+  - AlwaysAllow authorizer
+    threat: it by-pass authorizers
+  - Admission control:
+    - AlwaysAllow
+      threat: it by-pass admission controls
+    - NamespaceAutoProvision
+      threat: it allows accidental namespace creation
+  - profiling
+
+  The following are required:
+
+  - RBAC authorizer
```

Comment:
I don't think we should require RBAC. We're already testing for authorization behavior, as long as that's enforced, do we care about the mechanism?